### PR TITLE
NeRF Implementation

### PR DESCRIPTION
**Pull request by YanivHollander**
#### Changes
A fully fledged implementation of NeRF designed for Kornia

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Review by ducha-aiki — APPROVED** [review]
LGTM as soon as it pass tests. Right now it seems there is a problem with `grid_apply`

**Comment by YanivHollander** [comment]
The function `grid_apply` is not part of my contribution. Is the problem there fixed by another PR maybe?

**Comment by ducha-aiki** [comment]
@YanivHollander True, no it fails with 
image

**Comment by YanivHollander** [comment]
OK, I think this commit should fix the error: https://github.com/kornia/kornia/pull/1911/commits/7be3674cb2e22bd0bc6b9525a82373b91885db3f

**Comment by ducha-aiki** [comment]
@YanivHollander  this is what I (mistakenly) called "grid_sample":

```

kornia/nerf/rays.py:379: in calc_ray_params
[17143](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17144)
    points_2d_camera = self.sample_points_2d(cameras.height, cameras.width)
[17144](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17145)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[17145](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17146)

[17146](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17147)
self = <kornia.nerf.rays.UniformRaySampler object at 0x7f44ebcc29a0>
[17147](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17148)
heights = tensor([5, 4, 4, 4], dtype=torch.int32)
[17148](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17149)
widths = tensor([9, 7, 7, 7], dtype=torch.int32), sampling_step = 1
[17149](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17150)

[17150](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17151)
    def sample_points_2d(self, heights: Tensor, widths: Tensor, sampling_step=1) -> Dict[int, RaySampler.Points2D]:
[17151](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17152)
        r"""Uniformly sample pixel points in 2d for all scene camera pixels.
[17152](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17153)
    
[17153](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17154)
        Args:
[17154](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17155)
            heights: tensor that holds scene camera image heights (can vary between cameras): math: `(B)`.
[17155](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17156)
            widths: tensor that holds scene camera image widths (can vary between cameras): math: `(B)`.
[17156](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17157)
            sampling_step: defines uniform strides between rows and columns: int.
[17157](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17158)
    
[17158](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17159)
        Returns:
[17159](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17160)
            dictionary of Points2D objects that holds information on pixel 2d coordinates of each ray and the camera
[17160](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17161)
              id it was casted by: Dict[int, Points2D]
[17161](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17162)
        """
[17162](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17163)
        heights = heights.int()
[17163](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17164)
        widths = widths.int()
[17164](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17165)
        points2d_as_flat_tensors: Dict[int, RaySampler.Points2D_FlatTensors] = {}
[17165](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17166)
        for camera_id, (height, width) in enumerate(zip(heights.tolist(), widths.tolist())):
[17166](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17167)
            n = height * width
[17167](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17168)
>           y_grid, x_grid = torch.meshgrid(
[17168](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17169)
                torch.arange(0, height, sampling_step, device=self._device, dtype=self._dtype),
[17169](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17170)
                torch.arange(0, width, sampling_step, device=self._device, dtype=self._dtype),
[17170](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17171)
                indexing='ij',
[17171](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17172)
            )
[17172](https://github.com/kornia/kornia/actions/runs/3160702438/jobs/5152154461#step:5:17173)
E           TypeError: meshgrid() got an unexpected keyword argument 'indexing'
```
image

**Comment by YanivHollander** [comment]
Yes, @ducha-aiki, I saw these errors. As fr as I understand those are for non-required tests. The problem is that this argument was introduced in later versions of PyTorch. I saw it was also being used by other modules in Kornia. What do you suggest I should do?

**Comment by ducha-aiki** [comment]
@YanivHollander there is a compat function `torch_meshgrid` in kornia for exactly this case, I have done the fix.

**Comment by ducha-aiki** [comment]
@YanivHollander Thanks again for the terrific contribution! :)

Could you please also add the documentation into a separate PR?
Which means create `nerf.rst` here https://github.com/kornia/kornia/tree/master/docs/source and reference it from https://github.com/kornia/kornia/tree/master/docs/source/index.rst

**Comment by YanivHollander** [comment]
Thank you @ducha-aiki for merging my NeRF contribution into Kornia. I will proceed with what you requested regarding documentation. When is the next planned release date? I may have few more additions as I move along with new features to the current NeRF version.

**Comment by ducha-aiki** [comment]
@YanivHollander we plan this week, mostly because we planned to do bug-fixing release anyway.
Don't feel constrained by the release.

File: kornia/nerf/nerf_solver.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        if cameras.batch_size != len(imgs):
+            raise ValueError('Number of cameras must match number of input images')
+        if not all(img.shape[0] == 3 for img in imgs):
+            raise ValueError('All images must have three RGB channels')
+        if not all(height == img.shape[1] for height, img in zip(cameras.height.tolist(), imgs)):
+            raise ValueError('All image heights must match camera heights')
+        if not all(width == img.shape[2] for width, img in zip(cameras.width.tolist(), imgs)):
+            raise ValueError('All image widths must match camera widths')
+
+    def _train_one_epoch(self) -> float:
```
Comment (edgarriba) [inline kornia/nerf/nerf_solver.py:120]: can we target as one of next steps to update `NerfSolver` using the `kornia.Trainer` (and updating whatever is necessary) ?

File: kornia/nerf/positional_encoder.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    Args:
+        num_dims: Number of input dimensions (channels): int
+        num_freqs: Number of frequency bands for encoding span: int
+        log_space: Whether frequency sampling should be log spaced: bool
+    """
+
+    def __init__(self, num_dims: int, num_freqs: int, log_space: bool = False) -> None:
+        super().__init__()
+        self._num_dims = num_dims
+        self._embed_fns = [lambda x: x]
```
Comment (edgarriba) [inline kornia/nerf/positional_encoder.py:19]: have we tried to torchscript ? does it support well lambda ops?

File: kornia/nerf/rays.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+import torch
+
+from kornia.core import Device, Tensor
+from kornia.geometry.camera import PinholeCamera
+from kornia.nerf.camera_utils import cameras_for_ids
+from kornia.utils._compat import torch_meshgrid
+from kornia.utils.helpers import _torch_inverse_cast
+
+
+class RaySampler:
```
Comment (edgarriba) [inline kornia/nerf/rays.py:13]: as pointed above, some parts of this class could be abstracted as a generic Ray class

File: kornia/nerf/rays.py [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+    @property
+    def points_2d(self) -> Tensor:
+        return self._points_2d
+
+    def __len__(self) -> int:
+        if self.origins is None:
+            return 0
+        return self.origins.shape[0]
+
+    def _calc_ray_directions_cam(self, cameras: PinholeCamera, points_2d: Tensor):
```
Comment (edgarriba) [inline kornia/nerf/rays.py:57]: this can be a member function a futur Ray class

**Review by edgarriba — COMMENTED** [review]
Added a bunch of technical comments for next steps of this project. @YanivHollander @ducha-aiki I believe is worth to start an issue tracker about `kornia.nerf` and list them in terms of roadmap to give also visibility. @YanivHollander amazing contribution

**Comment by edgarriba** [comment]
forgot to add that docs were not included/rendered

**Comment by YanivHollander** [comment]
@edgarriba, to answer few of your questions above:

1. If you remember, I already had a number of Jupyter notebooks for testing various parts in this NeRF implementation, including one that runs the entire process. I'm thinking to convert this notebook to a tutorial
2. My next step, though, would be on the .rst doc files
3. I ran NeRF many times, but mostly focused on debuging and quality parameterization. I do not have runtime benchmarks yet, but will get to this later. I can say that the current implementation is indeed slow, and runs for the examples I tried for couple of hours on a Colab pro machine (which I believe runs one of the latest Nvidia RTX) 
4. Instant NeRF is surely on my radar...
5. I'm compiling a roadmap for future features that can be added by myself and others

**Comment by edgarriba** [comment]
@YanivHollander  awesome, thank you!

**Pull request merged**